### PR TITLE
Add expr_len parameter to Rust circuit drawer

### PR DESCRIPTION
**Pull request by KevesDev**
### AI/LLM disclosure

- [ ] I didn't use LLM tooling, or only used it privately.
- [ ] I used the following tool to help write this PR description: Claude Code (Sonnet 4.6)
- [x] I used the following tool to generate or modify code: Claude Code (Sonnet 4.6), reviewed and approved by me

---

## What does this PR do?

Adds an `expr_len` parameter to the Rust circuit drawer, closing the feature-parity gap with the Python text drawer identified in #16464. `expr_len` controls how many Unicode grapheme clusters are shown for classical `Expr` condition labels (e.g. `IfElseOp`/`SwitchCaseOp` conditions) before they are truncated with `"..."`. The parameter is threaded through the full rendering pipeline: `CircuitDrawerConfig` (C API struct), `qk_circuit_draw()` (default `30` when `config` is NULL), `draw_circuit()`, and `TextDrawer`. A grapheme-cluster-aware `truncate_to_expr_len()` helper is added and wired into a `TextDrawer::truncate_expr_label()` method, ready to be called from `get_label()` when PR #16063 adds control-flow op support to the Rust drawer.

## Why was this PR needed?

The Python text drawer has supported `expr_len` since PR #10869 (2023-10-18). The Rust drawer (`draw_circuit()` and `CircuitDrawerConfig`) was created in PR #15357 (2026-03-19) without this parameter. When the Rust drawer eventually replaces the Python text drawer, passing `expr_len` to `QuantumCircuit.draw()` would have had no effect on the Rust rendering path — the value could not be passed through to the rendering engine at all. Adding the parameter now (before control-flow ops land in the Rust drawer via #16063) means callers can already set `expr_len` on the C API side and the truncation will activate automatically once control-flow rendering is added, with no second pass needed.

The truncation uses `unicode_segmentation::UnicodeSegmentation::graphemes()` (already imported) rather than byte or scalar-value slicing, so multi-byte and combining-character sequences are counted as single display units — the same effective behavior as Python's `str[:n]` on normalized Unicode strings.

## What are the relevant issue numbers?

Part of #16464 (`expr_len` sub-task)

Note: I'm using "Part of" rather than "Closes" intentionally — #16464 is a multi-task tracking issue with other open sub-tasks. Closing it from this PR would orphan those remaining items.

## Does this PR meet the acceptance criteria?

- [x] Tests added for new/changed behavior (5 new unit tests for `truncate_to_expr_len` in `circuit_drawer.rs`)
- [x] All tests passing (27/27 circuit drawer tests pass; 5 new tests cover below-limit, at-boundary, exceeds-limit, `expr_len=0` edge case, and multi-byte grapheme cluster correctness)
- [x] Follows project style guide (no `rustfmt`/`clippy` warnings on the changed files)
- [x] No breaking changes introduced (`expr_len` is a new, additive field; existing C callers that initialize `CircuitDrawerConfig` without `expr_len` will need to add it, matching `{bundle_cregs, merge_wires, fold, expr_len}`)
- [x] Documentation updated (C API doc comment on `CircuitDrawerConfig.expr_len`; release note added at `releasenotes/notes/rust-drawer-expr-len-16464-a3c7f91d2e084b5c.yaml`)

### CODE REVIEW

**Comment by KevesDev** [comment]
Hi @eliarbel — thank you for opening and actively maintaining the #16464 tracking issue. I've implemented the `expr_len` sub-task: threaded `expr_len: usize` through the full rendering pipeline (`CircuitDrawerConfig`, `qk_circuit_draw()`, `draw_circuit()`, `TextDrawer`) with a default of 30 matching the Python text drawer, added a grapheme-cluster-aware `truncate_to_expr_len()` helper and `TextDrawer::truncate_expr_label()` method ready for when #16063 lands, and added 5 unit tests covering boundary conditions, the `expr_len=0` edge case, and multi-byte Unicode correctness. Would appreciate a review when you have time!

**Comment by eliarbel** [comment]
Thanks for working on this! We'll get to reviewing it soon.

File: crates/circuit/src/circuit_drawer.rs [hunk 1]
```diff
@@ -799,10 +800,31 @@ pub const Q_CL_CROSSED_WIRE: char = '╪';
...
 
+/// Truncate `text` to at most `max_chars` Unicode grapheme clusters, appending
+/// `"..."` if the string was shortened. Mirrors the Python text drawer's
+/// `expr_len` truncation in `text.py:1349–1350`.
+///
+/// Grapheme-cluster boundaries are used (via `unicode_segmentation`) rather
+/// than byte or scalar-value indices so that multi-byte characters and
+/// combining sequences are counted as single display units, matching the
+/// behaviour a reader would expect when looking at the rendered circuit.
+#[allow(dead_code)]
```
Comment (namanlp) [inline crates/circuit/src/circuit_drawer.rs:811]: Why would you allow `dead_code`?

**Comment by namanlp** [comment]
I am sorry, but this looks completely AI generated to me.

**Comment by KevesDev** [comment]
> I am sorry, but this looks completely AI generated to me.

Hi,

As students of the Codepath Capstone project, we are directly required to use Claude for our work, and the task is to contribute to open-source projects. That said, there is an extremely rigorous workflow we also have to follow, and both our work itself and each of those steps is watched and judged by the mentors of the program. That includes members from Anthropic and quite a number of pretty skilled higher-level engineers.

Aside from that, while I absolutely am new to this repo, I have been programming for some 20 years myself and I chose this specific repo because I am extremely interested in quantum computing as a whole. Being able to contribute to something of this stature is a big deal to me and I take it seriously.

I disclosed that Claude was using the correct PR template, and made sure it was allowed. If you have issues with the code I will happily try and fix those, because things like that can always happen (especially when working on a large codebase like this, with a multitude of contributions).

Regarding the #allow(dead_code): The function is only exercised by the 5 unit tests, not from any production call path. The production call site would be in get_label()'s control-flow arm, which is currently unimplemented!() pending PR #16063. Without the attribute, rustc would emit a dead_code warning that would fail CI. Once #16063 lands the attribute comes off.

I hope that helps!

**Comment by eliarbel** [comment]
@KevesDev thanks for taking a stab into that flag! At first glance this seems like a good start. 

Let's wait for #16063 to make a bit more progress, so that this PR can be based on it and use the `expr_len` parameter in functional code paths (instead of relying on `#[allow(dead_code)]`). 
I plan to get back to working on #16063, so please stay tuned.
